duckduckgo/BrowserServicesKit
# Add Phishing Detection Feature

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1204023833050360/1207976613228509/f
iOS PR: https://github.com/duckduckgo/iOS/pull/3336
macOS PR: https://github.com/duckduckgo/macos-browser/pull/3206
What kind of version bump will this require?: Major

**Optional**:

Tech Design URL:  https://app.asana.com/0/481882893211075/1207156899292810/f
CC: https://app.asana.com/0/481882893211075/1207220724600204/f

**Description**:
Implement Phishing Detection library to facilitate end-to-end phishing detection feature. Including:
1. Background data updates
2. API client for updating data
3. Embedded datasets
4. Detection logic
5. Event firing 



**Steps to test this PR**:
1. Build on macOS
2. Ensure signed in via use-login.duckduckgo.com (only available internally)
3. Quit the app
4. Visit https://privacy-test-pages.site/security/badware/phishing.html
5. Ensure error page is thrown
6. Click advanced
7. Click "Accept Risk"
8. Ensure page loads
9. Play around with navigations (back/forward, etc.)
10. Disable the feature in Settings>General
11. Try edge-case test pages:
 - https://bad.third-party.site/security/badware/phishing-iframe-loader.html
 - https://bad.third-party.site/security/badware/phishing-meta-redirect.html - currently not working
 - https://bad.third-party.site/security/badware/phishing-js-redirector-helper.html - currently not working

The last two failing test cases are to be addressed in follow-up work. Currently this feature is hidden behind a feature-flag for internal-only builds.



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
